### ISSUE TITLE
Graphflood with weird flow patterns

### ISSUE THREAD

**Issue by wschwanghart** [TopoToolbox/libtopotoolbox#149]
Hi @bgailleton 
I ran some code to test graphflood in python:
```
import topotoolbox as tt
DEM = tt.load_dem('taiwan')
H = tt.run_graphflood(DEM,p = 10e-05)
H.show()
```
The resulting image should show the water depth which takes values of up to >5000 m. Any ideas what I am doing wrong here?
![2024-11-23_153851](https://github.com/user-attachments/assets/02d4ae15-1487-4023-8a7f-df6b29961d81)

**Comment by wkearn** [comment]
I believe this is a result of the boundary conditions for this particular DEM. If you run graphflood with some of the other examples (Tibet, Big Tujunga, perfectworld) or on a subset of the Taiwan DEM, the answers are more reasonable. The same problem occurs with the Taal Volcano DEM.

The default `bcs` argument to graphflood sets the edges of the grid as boundaries where flow can leave the domain and the interior pixels as regular pixels through which flow can be routed. For the whole Taiwan DEM, the edges are NaNs, and the boundaries that really matter are the interior coastal pixels.

However, I am not sure what boundary conditions should be applied in this case through the `bcs` argument to `run_graphflood`. @bgailleton probably has a better idea.

**Comment by bgailleton** [comment]
Hi Wolfgang and Will,
Will is right, at the moment graphflood has to apply priority flood to fill
local minimas with water.
Because of the physical nature of the model, I need a robust way to manage
boundary conditions and tell the model where flow can out/go/enter/leave...
There is no documentation yet and I'll need to at least add examples, but
in your case you need to feed bcs with a 2d matrix of dem dimension and of
type uint8.
0 is no data
1 is normal node
3 means flow can leave the model from there (and also it will be a starting
point of priority flood)
Other codes are more specific and non relevant here.

Long story short you need to set bcs where data is nan to 0, all the other
nodes to 1, and 3 at the outlet of the basins. A simple way to do that is
to set any node with data but bordered by a nan (or a bc ==0) to 3.

I have python routines to automate many different boundary nasks, I'll port
some to the c library when I'll have time

Let me know if that's clear, I'd send a code snippet but I am away from
computers for the weekend!

Cheers

**Comment by bgailleton** [comment]
Definitly! I won't have time to implement that soon, but in the meantime here is a `python` workaround for @wschwanghart:

```
import numpy as np
from scipy.ndimage import binary_dilation
import topotoolbox as tt
import matplotlib.pyplot as plt

DEM = tt.load_dem('taiwan').fillsinks()

# Step 1: Initialize bcs to 1
bcs = np.ones_like(DEM.z, dtype=np.uint8)

# Step 2: Set bcs to 0 where Z is NaN
bcs[np.isnan(DEM.z)] = 0

# Step 3: Set bcs to 3 where a node 1 is bordered by a 0
# Create a binary mask for `bcs == 0` and dilate it to find neighbors
bordering_zero_mask = binary_dilation(bcs == 0)
bcs[(bcs == 1) & bordering_zero_mask] = 3

# Step 4: Set bcs to 3 for any remaining 1s on the edge
rows, cols = bcs.shape
edges = np.zeros_like(bcs, dtype=bool)
edges[0, :] = edges[-1, :] = edges[:, 0] = edges[:, -1] = True  # Mark edges
bcs[(bcs == 1) & edges] = 3

H = tt.run_graphflood(DEM, p = 10e-05, bcs = bcs)

H.show()
```

You'll need `scipy` installed (`pip install scipy`) if not. Also note that I pre-filled the DEM, because `graphflood` gets unstable if you have really big local minimas/lakes (the friction equation reaches its limit in this case).

Also, not that the function may be callable from MATLAB (@wkearn ?) and otherwise you can save `mat` file from python using [`savemat`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.io.savemat.html)

**Comment by wkearn** [comment]
We don't have a wrapper for Graphflood in MATLAB yet, but it shouldn't be too hard to add -- I'll put it on my TODO list. 

Good to know that `scipy` provides morphological image processing routines. Morphological dilation is how we would normally do something like this in MATLAB using the Image Processing Toolbox, and it might be good to recommend `scipy` for functions that are not yet available in libtopotoolbox.

**Comment by wschwanghart** [comment]
Having it in MATLAB would be perfect. While it runs smoothly in Python,
scrutinizing and visualizing the results is easier in MATLAB, at least for
me.

Von Gmail Mobile gesendet


William Kearney ***@***.***> schrieb am Mo. 25. Nov. 2024 um
14:14:

**Comment by bgailleton** [comment]
`graphflood` in libtopotoolbox does not rely on any particular `python`-specific things, as long as you can provide the 2D arrays in the given data types (although I prefer `python` on my side for data analysis but that's not a surprise).

Just note that the GPU version, substantially faster, is `python`-only. I don't think I would be able to port it to MATLAB or C (or it would need a LOT of work, especially to make it portable)

**Comment by bgailleton** [comment]
Thanks a lot :D 

Yeah it would be even more straightforward than before as I streamlined the install of my tools (`pip install pyscabbard`)

**Comment by wschwanghart** [comment]
I think we can close this issue, right?

**Comment by bgailleton** [comment]
Yes it should be fixed by the new graphflood's update on PR #203
